File: examples/python/example_sdp.py
import clarabel;
import numpy as np;
from scipy import sparse;

# Define problem data
P = sparse.eye(6).tocsc()
A = sparse.eye(6).tocsc()

c = np.zeros(6);
b = np.array([-3., 1., 4., 1., 2., 5.]);

cones = [clarabel.PSDTriangleConeT(3)]
settings = clarabel.DefaultSettings();

solver = clarabel.DefaultSolver(P,c,A,b,cones,settings);
solution = solver.solve()
print(
    f"Solver terminated with status: {solution.status}, objective {solution.obj_val},\n"
    f"and solution: {dict(s=solution.s, x=solution.x, z=solution.z)}"
)

File: examples/rust/example_sdp.rs
#![allow(non_snake_case)]

fn main() {
    use clarabel::algebra::*;
    use clarabel::solver::*;

    let n = 3;
    let nvec = (n * (n + 1)) >> 1;

    // SDP Example
    let P = CscMatrix::zeros((nvec, nvec));
    let c = vec![1., 0., 1., 0., 0., 1.];

    let sqrt2 = 2f64.sqrt();
    let A = CscMatrix::from(&[
        [-1., 0., 0., 0., 0., 0.],
        [0., -sqrt2, 0., 0., 0., 0.],
        [0., 0., -1., 0., 0., 0.],
        [0., 0., 0., -sqrt2, 0., 0.],
        [0., 0., 0., 0., -sqrt2, 0.],
        [0., 0., 0., 0., 0., -1.],
        [1., 4., 3., 8., 10., 6.],
    ]);

    let mut b = vec![0.0; 6];
    b.push(1.);

    let cones = vec![PSDTriangleConeT(n), ZeroConeT(1)];

    let settings = DefaultSettings::default();

    let mut solver = DefaultSolver::new(&P, &c, &A, &b, &cones, settings);

    solver.solve();
}
